vm/compiler/src/coinbase_puzzle/mod.rs

```diff
@@ -0,0 +1,363 @@
+            Self::Verifier(_) => bail!("Cannot prove the coinbase puzzle with a verifier"),
+        };
+
+        let polynomial = Self::prover_polynomial(epoch_challenge, address, nonce)?;
+
+        let product_evaluations = {
+            let polynomial_evaluations = pk.product_domain.in_order_fft_with_pc(&polynomial, &pk.fft_precomputation);
+            let product_evaluations = pk.product_domain.mul_polynomials_in_evaluation_domain(
+                &polynomial_evaluations,
+                &epoch_challenge.epoch_polynomial_evaluations().evaluations,
+            );
+            product_evaluations
+        };
+        let (commitment, _rand) =
+            KZG10::commit_lagrange(&pk.lagrange_basis(), &product_evaluations, None, &AtomicBool::default(), None)?;
+        let point = hash_commitment(&commitment)?;
+        let product_eval_at_point = polynomial.evaluate(point) * epoch_challenge.epoch_polynomial().evaluate(point);
+
+        let proof = KZG10::open_lagrange(
+            &pk.lagrange_basis(),
+            pk.product_domain_elements(),
+            &product_evaluations,
+            point,
+            product_eval_at_point,
+        )?;
+        ensure!(!proof.is_hiding(), "The prover solution must contain a non-hiding proof");
+
+        debug_assert!(KZG10::check(&pk.verifying_key, &commitment, point, product_eval_at_point, &proof)?);
+
+        Ok(ProverSolution::new(PartialSolution::new(*address, nonce, commitment), proof))
+    }
+
+    /// Returns a coinbase solution for the given epoch challenge and prover solutions.
+    ///
+    /// # Note
+    /// This method does *not* check that the prover solutions are valid.
+    pub fn accumulate(
+        &self,
+        epoch_challenge: &EpochChallenge<N>,
+        prover_solutions: &[ProverSolution<N>],
+    ) -> Result<CoinbaseSolution<N>> {
+        // Ensure there exists prover solutions.
+        if prover_solutions.is_empty() {
+            bail!("Cannot accumulate an empty list of prover solutions.");
+        }
+
+        // Ensure the number of prover solutions does not exceed `MAX_PROVER_SOLUTIONS`.
+        if prover_solutions.len() > MAX_PROVER_SOLUTIONS {
+            bail!(
+                "Cannot accumulate beyond {MAX_PROVER_SOLUTIONS} prover solutions, found {}.",
+                prover_solutions.len()
+            );
+        }
+
+        // Retrieve the coinbase proving key.
+        let pk = match self {
+            Self::Prover(coinbase_proving_key) => coinbase_proving_key,
+            Self::Verifier(_) => bail!("Cannot accumulate the coinbase puzzle with a verifier"),
+        };
+
+        let (prover_polynomials, partial_solutions): (Vec<_>, Vec<_>) = cfg_iter!(prover_solutions)
+            .filter_map(|solution| {
+                if solution.proof().is_hiding() {
+                    return None;
+                }
+                let polynomial = solution.to_prover_polynomial(epoch_challenge).unwrap();
```

Comment:
is this `unwrap` guaranteed to not trigger?

Reply:
Good catch. It is potentially triggerable; the current limitation is because there's no `try_filter_map` in either `rayon` or `itertools`.